File: backend/app/schemas/scheduled_report.py
"""Pydantic schemas for scheduled-report CRUD (`/api/analytics/scheduled-reports`).

Every field the runner reads is validated here, against the runner's OWN
registries rather than a restated copy:

* ``report_type`` against ``services/report_export.EXPORTERS`` — the dict
  ``_generate_report_payload`` looks the exporter up in. A row naming a key that
  isn't there raises ``ValueError`` on every tick and burns through the 5-strike
  auto-disable without ever sending anything.
* ``cadence`` against ``services/scheduled_reports._CADENCE_DELTA`` — the runner
  falls back to daily on an unknown cadence, so an unvalidated value silently
  reschedules a "monthly" report as daily.

Importing the registries means adding a report type or a cadence updates this
surface for free; there is no second list to remember.
"""

from __future__ import annotations

import re
import uuid
from datetime import UTC, datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.services.report_export import EXPORTERS
from app.services.scheduled_reports import known_cadences

#: Valid values, derived from the runner's registries (see the module docstring).
REPORT_TYPES: tuple[str, ...] = tuple(sorted(EXPORTERS))
CADENCES: tuple[str, ...] = tuple(sorted(known_cadences()))

#: Same conservative shape check `api/signup.py` and `api/partner.py` use — we
#: don't pull in the `email-validator` dependency for this. Non-dot character
#: classes delimited by literal dots so the engine can't backtrack
#: catastrophically on an adversarial input.
_EMAIL_PATTERN = re.compile(r"^[^\s@]+@[^\s@.]+(?:\.[^\s@.]+)+$")

#: A schedule is a recurring outbound email carrying a CSV of the tenant's AP
#: spend, so the recipient list is a distribution surface, not a preference.
#: Bounded on both ends: at least one address (a schedule with none can only
#: fail — the runner marks `failure` on every tick), and few enough that one row
#: can't be turned into a mailing blast.
MAX_RECIPIENTS = 20
MAX_PERIOD_DAYS = 366


def _clean_recipients(value: Any) -> list[str]:
    if not isinstance(value, list):
        raise ValueError("recipients must be a list of email addresses")
    # Cap the RAW list before doing any per-item work, so an oversized payload
    # is refused rather than parsed. (De-duping first would also let 5000
    # copies of one address slip under the cap.)
    if len(value) > MAX_RECIPIENTS:
        raise ValueError(f"at most {MAX_RECIPIENTS} recipients are allowed")
    cleaned: list[str] = []
    seen: set[str] = set()
    for raw in value:
        if not isinstance(raw, str):
            raise ValueError("recipients must be a list of email addresses")
        address = raw.strip()
        if not _EMAIL_PATTERN.match(address):
            # Never echo the offending value — a recipient list is third-party
            # PII and this message reaches an HTTP error body.
            raise ValueError("recipients must all be valid email addresses")
        key = address.lower()
        if key in seen:
            continue  # de-dupe rather than 422 — a duplicate would double-send
        seen.add(key)
        cleaned.append(address)
    if not cleaned:
        raise ValueError("at least one recipient is required")
    return cleaned


class ScheduledReportCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    report_type: str
    cadence: str
    recipients: list[str]
    period_days: int = Field(default=30, ge=1, le=MAX_PERIOD_DAYS)
    enabled: bool = True
    #: The first slot. Omitted → now, so the schedule runs on the next tick and
    #: the operator can see it work. `advance_next_run` then HOLDS whatever
    #: time-of-day this lands on, so pass an explicit value to pin e.g. 09:00.
    #: A value in the past is legitimate (it is immediately due, then catches up
    #: in whole cadence steps) — deliberately not rejected.
    next_run_at: datetime | None = None

    @field_validator("report_type")
    @classmethod
    def _known_report_type(cls, v: str) -> str:
        if v not in EXPORTERS:
            raise ValueError(f"report_type must be one of: {', '.join(REPORT_TYPES)}")
        return v

    @field_validator("cadence")
    @classmethod
    def _known_cadence(cls, v: str) -> str:
        if v not in CADENCES:
            raise ValueError(f"cadence must be one of: {', '.join(CADENCES)}")
        return v

    @field_validator("recipients")
    @classmethod
    def _valid_recipients(cls, v: Any) -> list[str]:
        return _clean_recipients(v)

    @field_validator("next_run_at")
    @classmethod
    def _aware(cls, v: datetime | None) -> datetime | None:
        # The column is `DateTime(timezone=True)`; a naive value means UTC, not
        # the server's local zone (see app/utils/dates.py).
        if v is not None and v.tzinfo is None:
            return v.replace(tzinfo=UTC)
        return v


class ScheduledReportUpdate(BaseModel):
    """Partial update — only the supplied fields change."""

    name: str | None = Field(default=None, min_length=1, max_length=120)
    report_type: str | None = None
    cadence: str | None = None
    recipients: list[str] | None = None
    period_days: int | None = Field(default=None, ge=1, le=MAX_PERIOD_DAYS)
    enabled: bool | None = None
    next_run_at: datetime | None = None

    @field_validator("report_type")
    @classmethod
    def _known_report_type(cls, v: str | None) -> str | None:
        if v is not None and v not in EXPORTERS:
            raise ValueError(f"report_type must be one of: {', '.join(REPORT_TYPES)}")
        return v

    @field_validator("cadence")
    @classmethod
    def _known_cadence(cls, v: str | None) -> str | None:
        if v is not None and v not in CADENCES:
            raise ValueError(f"cadence must be one of: {', '.join(CADENCES)}")
        return v

    @field_validator("recipients")
    @classmethod
    def _valid_recipients(cls, v: Any) -> list[str] | None:
        if v is None:
            return None
        return _clean_recipients(v)

    @field_validator("next_run_at")
    @classmethod
    def _aware(cls, v: datetime | None) -> datetime | None:
        if v is not None and v.tzinfo is None:
            return v.replace(tzinfo=UTC)
        return v


class ScheduledReportResponse(BaseModel):
    id: uuid.UUID
    name: str
    report_type: str
    cadence: str
    recipients: list[str]
    period_days: int
    enabled: bool
    next_run_at: datetime
    last_run_at: datetime | None = None
    last_run_status: str | None = None
    last_run_error: str | None = None

    @classmethod
    def from_db(cls, row) -> ScheduledReportResponse:
        return cls(
            id=row.id,
            name=row.name,
            report_type=row.report_type,
            cadence=row.cadence,
            recipients=list(row.recipients or []),
            period_days=row.period_days,
            enabled=row.enabled,
            next_run_at=row.next_run_at,
            last_run_at=row.last_run_at,
            last_run_status=row.last_run_status,
            last_run_error=row.last_run_error,
        )


class ScheduledReportListResponse(BaseModel):
    schedules: list[ScheduledReportResponse]
    #: The catalogs the client picks from, so a UI never hardcodes them.
    report_types: list[str] = Field(default_factory=lambda: list(REPORT_TYPES))
    cadences: list[str] = Field(default_factory=lambda: list(CADENCES))

File: frontend/src/lib/types/scheduledReport.ts
/**
 * Types for the scheduled-report admin surface (`/cfo` → Scheduled Reports).
 *
 * Mirrors the backend `ScheduledReportResponse` / list envelope under
 * `/api/analytics/scheduled-reports`. The runner
 * (`backend/app/services/scheduled_reports.py`) shipped complete and this CRUD
 * surface did not, so a schedule could only be inserted by hand-written SQL and
 * `list_due_schedules` returned `[]` on every tick forever.
 *
 * Two deliberate shape choices:
 *
 *  - `report_type` and `cadence` are **plain strings, not unions**. Both
 *    vocabularies come off the runner's own registries and ride the list
 *    envelope (`report_types` / `cadences`); a TypeScript union here would turn
 *    "the backend gained a report type" into a compile error on an unchanged
 *    frontend and — worse — invite a hardcoded copy of the list. The UI drives
 *    both selects off the response and humanises an unknown key rather than
 *    hiding it (see `components/analytics/scheduledReportDisplay.ts`).
 *  - `last_run_status` IS a union: it is a fixed lifecycle vocabulary the UI
 *    branches on (a `failure` renders differently from a `partial`), so a new
 *    member is a real UI decision, not a data value to pass through.
 *
 * There is no `entity_id` and no entity selector anywhere on this surface:
 * `X-Entity-ID` is not honoured by the scheduled-report routes — a schedule is
 * whole-tenant by construction.
 */

/** Outcome of the runner's last attempt. `null` = never run yet. */
export type ScheduledReportRunStatus = 'success' | 'partial' | 'failure';

export interface ScheduledReport {
	id: string;
	name: string;
	/** A key from the list envelope's `report_types` — never a closed union. */
	report_type: string;
	/** A key from the list envelope's `cadences` — never a closed union. */
	cadence: string;
	/** De-duped case-insensitively by the backend, so this is the list that was
	 *  actually SAVED — which is not necessarily the list that was submitted. */
	recipients: string[];
	period_days: number;
	enabled: boolean;
	next_run_at: string;
	last_run_at: string | null;
	last_run_status: ScheduledReportRunStatus | null;
	/** Counts + an exception class only — never a recipient address. Shown
	 *  verbatim for `partial` (some recipients DID receive it) and `failure`. */
	last_run_error: string | null;
}

/** `GET /api/analytics/scheduled-reports` — the rows plus the two vocabularies
 *  that drive the create / edit selects. */
export interface ScheduledReportList {
	schedules: ScheduledReport[];
	report_types: string[];
	cadences: string[];
}

/** `POST` body. `period_days` / `enabled` / `next_run_at` are optional; an
 *  omitted `next_run_at` means "now", i.e. it fires on the next runner tick.
 *  Pass an explicit value to pin a time-of-day (a past value is legitimate). */
export interface ScheduledReportCreate {
	name: string;
	report_type: string;
	cadence: string;
	recipients: string[];
	period_days?: number;
	enabled?: boolean;
	next_run_at?: string;
}

/** `PATCH` body — any subset; only what is sent changes. */
export type ScheduledReportPatch = Partial<ScheduledReportCreate>;
